contrib/win32/win32compat/win32_usertoken_utils.c

```diff
@@ -395,6 +396,56 @@ get_user_token(const char* user, int impersonation) {
 	return token;
 }
 
+HANDLE
+get_current_user_token()
+{
+	HANDLE current_token = 0;
+	HANDLE primary_token = 0;
+	PHANDLE user_token = 0;
+	PHANDLE token_dup = 0;
+	PWTS_SESSION_INFO session_info = 0;
+	DWORD count = 0, i = 0;
+	int session_id = 0;
+	BOOL ret = FALSE;
+
+	// Get the list of all terminal sessions
+	WTSEnumerateSessions(WTS_CURRENT_SERVER_HANDLE, 0, 1,
+		&session_info, &count);
+
+	int data_size = sizeof(WTS_SESSION_INFO);
+
+	// look over obtained list in search of the active session
+	for (i = 0; i < count; ++i)
+	{
+		WTS_SESSION_INFO si = session_info[i];
+		if (WTSActive == si.State)
+		{
+			// If the current session is active, store its ID
+			session_id = si.SessionId;
+			break;
+		}
+	}
+
+	WTSFreeMemory(session_info);
+
+	// Get token of the logged in user by the active session ID
+	ret = WTSQueryUserToken(session_id, &current_token);
```

Comment:
Major comment.

windows 10, windows 11 allows multiple RDP sessions. https://www.helpwire.app/blog/allow-multiple-remote-desktop-connections-windows-10/

If a customer has multiple RDP sessions then line 432 doesn't guarantee to return the session ID associated with the correct login user? We end up creating the process in different user namespace?

Reply:
**This again leads to security issue.**

Let's say two users have active RDP sessions. One of them is non-admin and the other is admin. 
Let's say non-admin runs ssh-add with a malicious PKCS11 provider, this function returns the token associated with admin user then he gains the access to the whole SYSTEM?

Reply:
You are correct that `get_current_user_token()` assumes there is only single logged-in user. We can easily extend it to make sure that it returns the token of the logged-in user that the connection originated from and remove this assumption. I'll address that shortly. Thanks!

Reply:
With the latest commit, we use the `pid` of the client connecting to `ssh-agent` to retrieve the user token, and use that token to start the helper as that user. That way, even if there are multiple logged-on users, the helper will be spawned as the exact user who originated the request.
I've verified with procmon that the helper is spawned as the expected user.
Thanks for bringing up the multi logged-on users scenario!